docs/DafnyRef/Types.md

```diff
@@ -4495,4 +4495,65 @@ imposed on the body of `ReachableVia` makes sure that, if the
 predicate returns `true`, then every object reference in `p` is as old
 as some object reference in another parameter to the predicate.
 
+## 6.5. Nameonly Formal Parameters and Default-Value Expressions
+
+A formal parametes of a method, constructor in a class, iterator,
+function, or datatype constructor can be declared with an expression
+denoting a _default value_. This makes the parameter _optional_,
+as opposed to _required_. All required parameters must be declared
+before any optional parameters. 
+
+For example,
+<!-- %check-resolve %save f.tmp -->
+```dafny
+function f(x: int, y: int := 10): int
+```
+may be called as either
+<!-- %check-resolve %use f.tmp -->
+```dafny
+const i := f(1,2);
+const j := f(1);
+```
+where `f(1)` is equivalent to `f(1,10)` in this case.
+
+Formal parameters may also be declared `nameonly`, in which case a call site
+must explicitly name the formal when providing its actual argument.
+
+The above function may be called as
+<!-- %no-check -->
+```dafny
+var k := f(y:=10, x:=2);
+```
+using names; here, though there are no `nameonly` parameters so the names may be used 
+or the actual arguments may be given in order.
+
+If a function is declared with a `nameonly` formal, then that formal's value must be given with a named assignment.
+For example, a function `ff` declared as
+<!-- %check-resolve -->
+```dafny
+function ff(x: int, nameonly y: int): int
+```
+may be called only using `ff(0, y := 4)`. A `nameonly` formal may also have a default value and thus be optional.
+
+Any formals after a `nameonly` formal must either be `nameonly` themselves or have default values.
+Otherwise an error will result when an attempt is made to call the function or method with such a formal list.
+
+The formals of datatype constructors are not required to have names. Such formals may have default values but
+may not be declared `nameonly`. Such nameless formals must be declared before `nameonly` formals.
```

Comment:
No, a nameless formal is neither allowed to have a default value nor allowed to be declared `nameonly`. Also, missing here is the rule that a nameless formal parameter is not allowed to follow an optional parameter. I suggest replacing the second and third sentences with "A nameless parameter is not allowed to follow a parameter with a default value.".

Reply:
OK, but I would think a default value for a nameless formal to be a perfectly reasonable thing.

Reply:
I agree. It would be a perfectly reasonable thing. We could add it in the future.